### ISSUE TITLE
Knora API Server release process

### ISSUE THREAD

**Issue by subotic** [dhlab-basel/Knora#522]
Some time ago, there was an email exchange regarding the release versioning scheme we would like to use. As far as I remember, the consensus was to use https://semver.org.

To make the release process as painless as possible, it would be nice to automate it as possible. Something along these lines: http://blog.byjean.eu/2015/07/10/painless-release-with-sbt.html

**Comment by subotic** [comment]
@lrosenth @benjamingeer @tobiasschweizer @mrivoal  

I would suggest the following release numbering:

- v1.1.0
- v1.2.0
- v1.3.0
...

I would like to create milestones with corresponding names. This way, we can each issue the corresponding milestone. As soon as all issues for this milestone are closed, we can tag `develop` with the corresponding version.

@loicjaouen would something like this help you?

**Comment by loicjaouen** [comment]
it sounds great, it would sure help.

It also sounds compatible with the [git standards](https://help.github.com/articles/creating-releases/).

If I understand well, the cycle goes like this:

1. add scripts needed to move an installation from the previous release
1. tag/release `develop` (f.e. `release-1.0.0`)
1. upon development needed on a release
   - create a branch from that tag (f.e. `fix/thisbug`)
   - code
   - test manually
1. tag `fix/thisbug` as `release-1.0.1`
1. maybe merge `fix/thisbug` into `develop` for further releases
  one might pick only the test cases if the fix touches some code that has diverged too much

(automatic build and test would even be nicer)

**Comment by subotic** [comment]
I'm not sure how the bugfix releases will be managed. The idea is, that since everything is going to be merged back into `develop`, we only have tags on the `develop` branch. If in the meantime a new feature gets merged into `develop` and the bugfix afterward, then we make a combined release.

If we start to make long-lived branches with only bugfix releases, then we are going to have a big mess pretty soon.

`develop` should always work, since all our tests are always run. We could expand it further with your test data so that you are also sure that it works for you.

**Comment by loicjaouen** [comment]
release management is generally a mess ;)

if we tag `release-1.0.0` and then `release-2.0.0` we then find a bug in `release-1.0.0`, it will be hard to fix it in `develop` that is already past `2.0.0` and still make the releases `1.0.1` and `2.0.1` from `develop`.

Let's start any way and see how it goes.

**Comment by subotic** [comment]
I don't think that we are going to have v1.0.0 and v2.0.0 at the same time. I was thinking more of a linear release management. The highest number is the latest release, so if v1.0.0 has a bug that needs fixing, then v2.0.0 either has the fix already or will also need the bugfix. In this case, we would have v2.0.1.

There seems to be a confusion with v1 and v2 of the API and the numbering of the releases.

At the moment, we have v1 and v2 routes in the API. In future, we are probably going to have v3, v4, etc. All API versions need to run in parallel and at the same time since we need to support projects over the long-term.

This means that, for example, a release tagged with v1.1.0 will have version 1 (production) and version 2 (alpha) of the API.

According to https://semver.org, we should make a major release when breaking changes are introduced. Since we cannot introduce breaking changes because of the long-term support we need to provide, we introduced API versions.

```
release                             api version
-------                             -----------

v1.1.0                              v1 (production) and v2 (alpha)
...
v2.0.0                              v1 (production) and v2 (production)
```

**Comment by musicEnfanthen** [comment]
Would something like [git flow](https://www.atlassian.com/git/tutorials/comparing-workflows/gitflow-workflow) fullfill the necessities? There is a pretty straigthforward separation of bugfixes, fetaures, releases etc.

**Comment by loicjaouen** [comment]
@subotic : sorry about that, it is all clear and I should have take an example with different numbers ;)

**Comment by loicjaouen** [comment]
so let's tag api version numbers as vX and releases as rX.X.X ?

**Comment by subotic** [comment]
@musicEnfanthen That was my initial idea. That's why we have `develop` and `master` branches. The only difference is that we don't use the `master`branch. But there is no reason why we couldn't start to use it. Basically, we could merge every tagged release from `develop` into master.

@loicjaouen I would only tag releases. With tagging, I mean to create a git tag. There the convention is vX.X.X.

The API version numbers will be present in the release notes and the documentation. The API version numbers don't really matter in the context of the releases since every release is going to include all available API versions. We could introduce the convention that we increment the major version number whenever an API version becomes stable. This means, that for releases with v.1.x.x only `v1` of the API can be assumed as stable and should be used in production. 

What we need to do for sure, is prominently write in the release notes what the status of each API version is, e.g, pre-alpha, alpha, beta, release candidate, stable. Also, somewhere we need to define what pre-alpha, alpha, beta, release candidate, stable actually mean.

**Comment by musicEnfanthen** [comment]
> Basically, we could merge every tagged release from develop into master.

Yes, as I understand this is also the main principle behind `git flow`. Finished features get pushed into `develop`, and only the (tagged) releases will go into `master` (beside the hotfixes that will immediately go to `master` and `develop` ). So you have always a working `develop` branch, and a stable `master` (release) branch.

**Comment by subotic** [comment]
I had a discussion with @lrosenth about version numbering. We came to the conclusion that it would be clearer if we follow semantic versioning as it is meant.

Here is the excerpt from http://semver.org:

> Given a version number MAJOR.MINOR.PATCH, increment the:
>
> - MAJOR version when you make incompatible API changes,
> - MINOR version when you add functionality in a backwards-compatible manner, and
> - PATCH version when you make backwards-compatible bug fixes.

The major change to the way we number our releases now, would be to also increment the MAJOR number in the case when any kind of changes to existing data would be necessary, e.g., any changes to the `Knora-Base` ontologies which are not backwards compatible.

We think that this would comunicate the need for data migration much clearer, since now they wouldn't need to read the release notes to know if they can update without the need for data migration.

**Comment by subotic** [comment]
This Thursday is again release day. I will release `V2.0.0` which from then on will be released with the new version numbering schema. I will then also simplify the release notes, so that we have only one document per major release, e.g. `V2.x.x.md`, `V3.x.x.md`, etc. This should also make it easier for us.
